bpf/process/types/probe_read_kernel_or_user.h

```diff
@@ -0,0 +1,150 @@
+}
+#endif // __LARGE_BPF_PROG
+
+/*
+ * bpf_core_read_kernel_or_user() abstracts away bpf_probe_read_kernel_or_user() call and captures offset
+ * relocation for source address using __builtin_preserve_access_index()
+ * built-in, provided by Clang.
+ */
+#define bpf_core_read_kernel_or_user(userspace, dst, sz, sz_mask, src)                \
+	probe_read_kernel_or_user(dst, sz, sz_mask,                                   \
+				  (const void *)__builtin_preserve_access_index(src), \
+				  userspace)
+
+/*
+ * bpf_core_read_kernel_or_user_str() is a thin wrapper around bpf_probe_read_kernel_or_user_str()
+ * additionally emitting BPF CO-RE field relocation for specified source
+ * argument.
+ */
+#define bpf_core_read_kernel_or_user_str(userspace, dst, sz, src)                         \
+	probe_read_kernel_or_user_str(dst, sz,                                            \
+				      (const void *)__builtin_preserve_access_index(src), \
+				      userspace)
+
+/*
+ * BPF_CORE_READ_KERNEL_OR_USER_INTO() is a more performance-conscious variant of
+ * BPF_CORE_READ_KERNEL_OR_USER(), in which final field is read into user-provided storage.
+ * See BPF_CORE_READ_KERNEL_OR_USER() below for more details on general usage.
+ */
+#define BPF_CORE_READ_KERNEL_OR_USER_INTO(userspace, dst, src, a, ...)         \
+	({                                                                     \
+		typeof(dst) dst_x = dst;                                       \
+		typeof(src) src_x = src;                                       \
+		typeof(a) a_x = a;                                             \
+		((userspace) ? (___core_read(bpf_core_read_user, dst_x,        \
+					     src_x, a_x, ##__VA_ARGS__))       \
+			     : (___core_read(bpf_core_read, dst_x, src_x, a_x, \
+					     ##__VA_ARGS__)))                  \
+	})
+
+/*
+ * BPF_CORE_READ_KERNEL_OR_USER_STR_INTO() does same "pointer chasing" as
+ * BPF_CORE_READ_KERNEL_OR_USER_STR() for intermediate pointers, but then executes (and returns
+ * corresponding error code) bpf_core_read_kernel_or_user_str() for final string read.
+ */
+#define BPF_CORE_READ_KERNEL_OR_USER_STR_INTO(userspace, dst, src, a, ...)                                       \
+	({                                                                                                       \
+		typeof(dst) dst_x = dst;                                                                         \
+		typeof(src) src_x = src;                                                                         \
+		typeof(a) a_x = a;                                                                               \
+		if (userspace)                                                                                   \
+			___core_read(bpf_core_read_user_str, dst_x, src_x,                                       \
+				     a_x, ##__VA_ARGS__) else ___core_read(bpf_core_read_str, dst_x, src_x, a_x, \
+									   ##__VA_ARGS__)                        \
+	})
```

Comment:
What happens if `userspace` is false?

Reply:
There is an else clause that calls bpf_core_read_str(). I guess 'make format' reorganised this so that it was less obvious.